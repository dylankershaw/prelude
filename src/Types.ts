export interface Song {
  master_metadata_album_artist_name: string;
  reason_end: string;
  ts: string;
}

export interface ChartData {
  [date: string]: number;
}