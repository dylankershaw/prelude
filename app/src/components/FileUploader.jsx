import React, { useCallback } from 'react';
import { useDropzone } from 'react-dropzone';

export default ({ setData }) => {
  const cleanData = useCallback(str => {
    const parsedData = JSON.parse(`[${str.replace(/\n/g, ',')}]`);
    
    const filteredData = parsedData.filter(
      s => s.master_metadata_album_artist_name && s.master_metadata_album_artist_name.includes('Above \u0026 Beyond')
    );
    
    return filteredData.reduce((acc, cur) => {
      const date = cur.ts.split(' ')[0];
      return { ...acc, [date]: (acc[date] || 0) + 1 };
    }, {});
  }, []);

  const onDrop = useCallback(acceptedFiles => {
    acceptedFiles.forEach(file => {
      const reader = new FileReader();

      reader.onabort = () => console.log('file reading was aborted');
      reader.onerror = () => console.log('file reading has failed');
      reader.onload = () => {
        const dataStr = reader.result;
        setData(cleanData(dataStr));
      };

      reader.readAsText(file);
    });
  }, []);

  const { getRootProps, getInputProps, isDragActive } = useDropzone({ onDrop });

  return (
    <div {...getRootProps()}>
      <input {...getInputProps()} />
      {isDragActive ? <p>Drop the files here ...</p> : <p>Drag 'n' drop some files here, or click to select files</p>}
    </div>
  );
};
