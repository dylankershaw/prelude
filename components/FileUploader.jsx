import React, {useCallback} from 'react';
import {useDropzone} from 'react-dropzone';

const FileUploader = ({setSongData}) => {
  const onDrop = useCallback(([file]) => {
    const reader = new FileReader();

    reader.onload = () => {
      const songs = JSON.parse(`[${reader.result.replace(/\n/g, ',')}]`);
      setSongData(songs);
    };

    reader.readAsText(file);
  }, []);

  const {getRootProps, getInputProps} = useDropzone({onDrop, accept: 'application/json, .json'});

  return (
    <div>
      <div {...getRootProps()}>
        <input {...getInputProps()} />
        <p>Click to upload (.json only)</p>
      </div>
    </div>
  );
};

export default FileUploader;
