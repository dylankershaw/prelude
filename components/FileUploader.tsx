import React, {useCallback} from 'react';
import {useDropzone} from 'react-dropzone';
import {Song} from '../Types';

import {breakpoints, colors} from '../theme';

interface Props {
  setSongData(songs: Array<Song>): void;
}

const FileUploader: React.FC<Props> = ({setSongData}) => {
  const onDrop = useCallback(([file]) => {
    const reader = new FileReader();

    reader.onload = () => {
      if (typeof reader.result !== 'string') return;
      const songs = JSON.parse(`[${reader.result.replace(/\n/g, ',')}]`);
      setSongData(songs);
    };

    reader.readAsText(file);
  }, []);

  const {getRootProps, getInputProps} = useDropzone({onDrop, accept: 'application/json, .json'});

  return (
    <div className='container'>
      <button {...getRootProps()}>
        <input {...getInputProps()} />
        <p>UPLOAD</p>
      </button>
      <p>(EndSong.json only please)</p>
      <style jsx>
        {`
          .container {
            display: flex;
            flex-direction: column;
            margin: 0 auto;
            width: fit-content;
          }

          @media screen and (min-width: ${breakpoints.sm}) .container {
            margin: initial;
          }

          button {
            background: ${colors.green};
            border: none;
            border-radius: 1rem;
            color: white;
            margin-top: 2rem;
            padding: 0.5rem;
          }
        `}
      </style>
    </div>
  );
};

export default FileUploader;
