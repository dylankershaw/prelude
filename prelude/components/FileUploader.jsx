import React, {useCallback, useState} from 'react';
import {useDropzone} from 'react-dropzone';

const FileUploader = () => {
  const [artist, setArtist] = useState('Above & Beyond');

  const onDrop = useCallback(
    ([file]) => {
      const reader = new FileReader();

      reader.onload = async () => {
        const resp = await fetch('/api/songs', {
          method: 'post',
          contentType: 'application/json',
          body: JSON.stringify({artist, data: reader.result})
        });

        handleResponse(resp);
      };

      reader.readAsText(file);
    },
    [artist]
  );

  const handleResponse = useCallback(() => {}, []);

  const {getRootProps, getInputProps} = useDropzone({onDrop, accept: 'application/json, .json'});

  return (
    <div>
      <input value={artist} onChange={e => setArtist(e.target.value)} />
      <div {...getRootProps()}>
        <input {...getInputProps()} />
        <p>Click to upload (.json only)</p>
      </div>
    </div>
  );
};

export default FileUploader;
