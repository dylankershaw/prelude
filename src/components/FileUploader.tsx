import React, {useCallback} from 'react';
import {Song} from '../Types';

import {breakpoints, colors} from '../theme';

interface Props {
  setSongData(songs: Array<Song>): void;
}

const FileUploader: React.FC<Props> = ({setSongData}) => {
  const handleDrop = useCallback(e => {
    const reader = new FileReader();

    reader.onload = () => {
      if (typeof reader.result !== 'string') return;

      const jsonIsLineDelimited = reader.result.indexOf('},') === -1;
      const songs = JSON.parse(
        jsonIsLineDelimited ? `[${reader.result.replace(/\n/g, ',')}]` : reader.result
      );

      setSongData(songs);
    };

    reader.readAsText(e.target.files[0]);
  }, []);

  return (
    <div className='container'>
      <label htmlFor='upload-input'>UPLOAD</label>
      <span className='small-text'>(EndSong.json only please)</span>
      <input id='upload-input' type='file' accept='application/json,.json' onChange={handleDrop} />
      <style jsx>
        {`
          .container {
            align-items: center;
            display: flex;
            flex-direction: column;
            margin: 2rem auto 0;
            width: fit-content;
          }
          
          @media screen and (min-width: ${breakpoints.sm}) .container {
            margin: initial;
          }

          input {
            background: ${colors.green};
            border: none;
            border-radius: 1rem;
            color: white;
            padding: 0.5rem;
          }

          .small-text {
            font-size: 0.8rem;
          }
        `}
      </style>
    </div>
  );
};

export default FileUploader;
