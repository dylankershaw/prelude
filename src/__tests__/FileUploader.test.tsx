import React from 'react';
import {render} from '@testing-library/react';

import FileUploader from '../components/FileUploader';
import {Song} from '../Types';

test('matches snapshot', () => {
  const {container} = render(<FileUploader setSongData={(s: Array<Song>) => {}} />);
  expect(container).toMatchSnapshot();
});
