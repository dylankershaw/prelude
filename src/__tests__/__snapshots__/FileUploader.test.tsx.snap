// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <button
      tabindex="0"
    >
      <input
        accept="application/json, .json"
        autocomplete="off"
        multiple=""
        style="display: none;"
        tabindex="-1"
        type="file"
      />
      <p>
        UPLOAD
      </p>
    </button>
    <p>
      (EndSong.json only please)
    </p>
    <style>
      
          .container {
            display: flex;
            flex-direction: column;
            margin: 0 auto;
            width: fit-content;
          }

          @media screen and (min-width: 48rem) .container {
            margin: initial;
          }

          button {
            background: #1DB954;
            border: none;
            border-radius: 1rem;
            color: white;
            margin-top: 2rem;
            padding: 0.5rem;
          }
        
    </style>
  </div>
</div>
`;
